[PR-4]Decision making module for automatic migration

AmitRoushan opened the pull request:
PR has following changes:
class for periodically watch metric and call prediction module for top migration candidate
class for predict kernel server node for replica.

***Note***: 
 1. PR is written on top of #746 and #753.  #746 and #753 are **closed** and  #794, #795, #796 are **opened** for handling metric management.
 2. This PR has addressed review comments for #745

Commit:
added migration decision making module

Commit:
review comment fix

quinton-hoole commented:
Thanks. Please rebase before requesting a review.

quinton-hoole commented:
Please check that the comments describing this PR are correct (all the PR's mentioned there have been closed, so I assume it's out of date).
And please confirm that all of this PR needs to be reviewed?  It's quite huge.
Then please self-review.
Then ask @VenuReddy2103 to review.
Then ask me to do a final review before merging.
Thanks.

AmitRoushan commented:
Review comment fix from PR #745:
1. General convention for package name is to use lowercase
    **Status**: "migrationDecision" module changed to "migrationdecision". 
2. Don't need logger for each instance of it. Make it private static final.
     **Status**: ```MetricWatcher``` logger changed to "private static final"
3. Why transient volatile ?
    **Status**: Removed "transient" modifier from ResettableTimer in ```MetricWatcher```
4. Instead if timer != null, then simply return can be more clear ?
   **Status**: Timer instance creation moved to constructor. So removed null check from start method of ```MetricWatcher```
5. What we need was nodeMetric from client to the kernelserver. But we took from kernelserver to the client. Did we assume both to be same ? They both are not necessary to be same.
   **Status**: Updated code as per suggestion. Now rate and latency are used from kernelClient node metric.

AmitRoushan commented:
@VenuReddy2103 , I have handled all review comments from PR #745. Please have a look.
@quinton-hoole   #746 and #753 are closed and new PR  #794, #795 and #796 are open respectively for microservice and node metric management. Same has been updated in PR description.

quinton-hoole commented:
@AmitRoushan I still cannot review this PR until #794 , #795 and #796 have been reviewed.  And I can't review those until https://github.com/amino-os/Amino.Run/pull/794#issuecomment-500074934 has been answered.  So basically all of these PR's are blocked.

Commit:
rebased with latest code

Commit:
corrected kvstore app

AmitRoushan commented:
@quinton-hoole 
I have reverted KVstore code changes. Changes are made to test migration.
Also we fixed one bug , fount during testing.
Please find the attached status report below:
- Build report 
![clean_build](https://user-images.githubusercontent.com/9818606/64438072-0c106500-d0e5-11e9-9abf-2885fd0267d3.png)

- KVStore App report 
![kvstore](https://user-images.githubusercontent.com/9818606/64438077-0fa3ec00-d0e5-11e9-94a4-712bc2abe507.png)

- Hankstodo App report 
![hankstodo](https://user-images.githubusercontent.com/9818606/64438081-116daf80-d0e5-11e9-97ee-562c53fcf934.png)

- MinnieTwitter App report 
![minnietwitter](https://user-images.githubusercontent.com/9818606/64438088-1468a000-d0e5-11e9-8f87-f36158e7d9c4.png)

- HankstodoRuby App report 
![Selection_027](https://user-images.githubusercontent.com/9818606/64438155-33ffc880-d0e5-11e9-9648-a1dd163e6942.png)

- KVstorejs App report
![Selection_028](https://user-images.githubusercontent.com/9818606/64438177-4548d500-d0e5-11e9-83ee-bac8d2fc1e94.png)

quinton-hoole commented on core/src/main/java/amino/run/kernel/server/KernelServerImpl.java:
```diff
@@ -483,6 +483,7 @@ public static void main(String args[]) {
 
             // Register against OMS
             ServerInfo srvInfo = createServerInfo(host, ksArgs.labels);
+            srvInfo.processorCount = Runtime.getRuntime().availableProcessors();
```
It's not clear to me why this initialization is not done in createServerInfo like all the other initialization?
Or better still, in the constructor of ServerInfo?
I'm uncomfortable with having an uninitialized processorCount in ServerInfo structure until someone remembers to initialize it.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+        } catch (Exception e) {
+            logger.warning("Failed to get replica dispatcher");
+            e.printStackTrace();
+            return false;
```
It seems it would be more appropriate to throw an exception here. 
What does the return value of this method actually mean?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service not available",
+                                replicaID));
+                e.printStackTrace();
```
Printing a stack trace to the console is generally not useful.  Rather log appropriately, or throw an exception (which is probably not appropriate in this case).

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+            } catch (MicroServiceNotFoundException e) {
+                logger.warning(
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service not available",
```
Lets improve this exception message. It's not the microservice that's unavailable.  It's not found in OMS.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+            } catch (MicroServiceReplicaNotFoundException e) {
+                logger.warning(
+                        String.format(
+                                "Skipping Replica [%s] evaluation as micro service replica not available",
```
As above.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+                    KernelServerNotFoundException {
+        /*
+           To measure time consumption by vertex Vk we use the following formula
+           tk = Tk,posk+i + summation { (Ski+Sik)/Rposk,posi + Lposk,posi }
```
Please explain what all these terms mean.  tK, Tk, posk+i, Ski, Sik, Rposk?  It's not at all clear what you mean here.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+                serverManager.getKernelServerMetric(replicaKernelServer);
+
+        // get metrics stat of RPC processing time from different kernel client
+        // TODO currently considering processing time from different kernel client for simplicity.
```
Please clarify.  It's not clear what you mean here.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+        // calculate time factor impact on RPC with latency and bandwidth
+        long averageExecutionTime = 0;
+        long summation = 0;
+        long data;
```
Please call this dataSize.  It's not data.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+
+        // calculate time factor impact on RPC with latency and bandwidth
+        long averageExecutionTime = 0;
+        long summation = 0;
```
This is not a useful variable name.   Do you mean the sum of the round trip times?  Perhaps call it rttSum?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+                microServiceManager.getMicroServiceMetric(replicaID);
+
+        // calculate time factor impact on RPC with latency and bandwidth
+        long averageExecutionTime = 0;
```
Needs to be renamed to avgElapsedTime to be consistent with replicaMetric.getValue().elapsedTime

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+
+            data = replicaMetric.getValue().dataSize;
+            nodeMetric = replicaNodeMetric.get(replicaMetric.getKey());
+            // kernel client utilized by micro service but not running in kernel server are
```
Under what condition does the client not run in a kernel server?  This doesn't seem to make sense?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+            averageExecutionTime = averageExecutionTime / replicaMetrics.size();
+        }
+
+        return averageExecutionTime + summation;
```
It's not clear to me that you can simply add the averageExecutionTime (which is actually the average elapsed time for RPC's) to the summation of all data transfer times and get a sensible result.  Surely you should either add an average to an average, or a summation to a summation?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+        return true;
+    }
+
+    private long aggregatedProcessingTime(ReplicaID replicaID)
```
It would be good to add a comment to describe what this method is intended to calculate.  Because it's not at all clear from the code.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+        return averageExecutionTime + summation;
+    }
+
+    private List<ReplicaID> getMigrationReplicas() {
```
Please add a comment describing what this method is intended to do.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -0,0 +1,282 @@
+                            String.format(
+                                    "MicroService [%s] not available. Skipping replicas in migration evaluation",
+                                    microServiceID));
+                    e.printStackTrace();
```
What is the purpose of printing stack traces all over the place like this?  It makes it very difficult to look at the console output and tell whether anything has gone wrong or not.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+                        new AbstractMap.SimpleEntry<InetSocketAddress, Long>(
+                                ks, getOptimizedTime(replicaID, currentKernelServer, ks)));
+            }
+        } catch (RemoteException e) {
```
Don;'t we need to throw these exceptions for the caller to handle, rather than just log them here?

quinton-hoole replied:
Oh, I see, if an exception occurs, you just return the current node as the best node.  Fair enough.  It seems better to catch this exception higher up the call stack, and explicitly avoid moving the replica in that code instead.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+        AbstractMap.SimpleEntry<InetSocketAddress, Long> bestTimeOptimization =
+                predictedTime.first();
+        // if best time optimization is less or equal to zero then return current kernel server
+        if (bestTimeOptimization.getValue() <= 0) {
```
This seems to indicate a bug in the code?  How is is possible to calculate a value less than 0? 
I think we'd do better throwing an exception here, or panicking.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+        }
+
+        AbstractMap.SimpleEntry<InetSocketAddress, Long> bestTimeOptimization =
+                predictedTime.first();
```
I think it would be worth mentioning in a comment that you're relying on this set being sorted by time here.  Sets are not usually sorted, so this is an unusual use of a set, and a bit confusing.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+            replicaMetrics = microServiceManager.getMicroServiceMetric(replicaID);
+        } catch (MicroServiceNotFoundException e) {
+            e.printStackTrace();
+            return 0;
```
This exception should be thrown.  This also answers my previous question above.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+            currentKernelServerMetrics = serverManager.getKernelServerMetric(currentKernelServer);
+            futureKernelServerMetrics = serverManager.getKernelServerMetric(futureKernelServer);
+        } catch (KernelServerNotFoundException e) {
+            e.printStackTrace();
```
Throw this exception.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+        return bestTimeOptimization.getKey();
+    }
+
+    private long getOptimizedTime(
```
Add a comment to explain what this method does.

quinton-hoole replied:
nit: It's not actually an optimization, it's a prediction of elapsed time on a different node.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+    private long getOptimizedTime(
+            ReplicaID replicaID,
+            InetSocketAddress currentKernelServer,
+            InetSocketAddress futureKernelServer)
```
nit: It's not actually a future kernel server.  It's a prediction related to an potential alternative kernel server.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+                                kernelClient,
+                                futureKernelServer,
+                                currentByteExchanged);
+            } catch (Exception e) {
```
We can't just catch Exceptions and ignore them like this.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+        long currentProcessingTime;
+        long currentByteExchanged;
+        long futureProcessingTime;
+        double currentIOTime;
```
Lets avoid all double precision arithmetic. It's unnecessary.  Just record times in a small enough unit (e.g. microseconds or nanoseconds) and just use integer arithmetic.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MigrationPredictor.java:
```diff
@@ -0,0 +1,218 @@
+                            "Node metric not available between kernel client [%s] and kernel server [%s]",
+                            kernelClient, KernelServer);
+            logger.warning(msg);
+            throw new Exception(msg);
```
Please throw a more specific exception type.  This is a not found exception.

quinton-hoole commented on core/src/main/java/amino/run/policy/DefaultPolicy.java:
```diff
@@ -158,6 +158,24 @@ public void onNotification(Notification notification) throws RemoteException {
                 /* Store these metrics to make them available for statistics and decision making module */
                 updateMetric(object.replicaId, object.metrics);
                 return;
+            } else if (notification instanceof MigrationNotification) {
```
I still don't understand why we're using notifications for all of these things. 
Surely we should just have a method to receive metrics, and another one to receive migration instructions?
What's the benefit of using Notifications instead?

quinton-hoole reviewed (changes requested):
See inline comments for requested changes.  Amit, if you are otherwise occupied, I can make the required changes.  Let me know.

Commit:
review comment fixes

AmitRoushan commented:
@quinton-hoole 
I have fixed most of review comments but still i need some more time to run some test scenario for migration.

quinton-hoole commented:
Thanks Amit

On Mon, Sep 9, 2019, 04:57 Amit Roushan <notifications@github.com> wrote:

> @quinton-hoole <https://github.com/quinton-hoole>
> I have fixed most of review comments but still i need some more time to
> run some test scenario for migration.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/amino-os/Amino.Run/pull/793?email_source=notifications&email_token=AKNAA6H25X3YXZABUOV67MLQIY2Y3A5CNFSM4HSZF2U2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD6HJGPY#issuecomment-529437503>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKNAA6AQPSFU42J76OXHWHDQIY2Y3ANCNFSM4HSZF2UQ>
> .
>

quinton-hoole commented on core/src/main/java/amino/run/kernel/common/ServerInfo.java:
```diff
@@ -17,8 +17,11 @@
+
     public ServerInfo(InetSocketAddress addr) {
         this.host = addr;
+        this.processorCount = Runtime.getRuntime().availableProcessors();
```
This will only work correctly when this constructor is called on the machine to which the server info refers.  Surely processorCount should be a parameter to the constructor, similar to addr?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -32,7 +39,9 @@
-    private TreeSet<AbstractMap.SimpleEntry<ReplicaID, Long>> replicaHeap;
+    private TreeSet<AbstractMap.SimpleEntry<ReplicaID, Long>>
+            replicaHeap; // should only used in getBestCandidates for selecting best microservice
+    // replica for migration
```
Nit - weird comment formatting.  Use /* */ instead?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -80,11 +89,28 @@ public void run() {
+                                                String.format(
+                                                        "Skipping Replica [%s] for migration as micro service not found",
+                                                        bestCandidateReplicaId));
+                                        e.printStackTrace();
```
Why do you need to print the stack trace?

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -80,11 +89,28 @@ public void run() {
+                                                String.format(
+                                                        "Skipping Replica [%s] for migration as micro service replica not found",
+                                                        bestCandidateReplicaId));
+                                        e.printStackTrace();
```
As above.

quinton-hoole commented on core/src/main/java/amino/run/oms/migrationdecision/MetricWatcher.java:
```diff
@@ -222,55 +276,75 @@ private long aggregatedProcessingTime(ReplicaID replicaID)
         }
 
-        return averageExecutionTime + summation;
+        return avgElapsedTime + avgRtt;
```
This seems wrong.  Don't you want to *subtract* the RTT from the elapsedTime to get the processing (computation) time?  Also, a single inbound RPC may result in multiple "dowstream" outbound RPC's, some in serial, and some in parallel.  Using the avgRtt here (even assuming your substract it as mentioned above) implies a single outbound RPC per inbound RPC, which is not correct?  I think you need to keep track of the ratio of inbound:outbound (say 1:n) and subtract n times the avgRtt?

quinton-hoole commented:
@AmitRoushan  How is your testing going?

quinton-hoole commented:
@AmitRoushan ?

Commit:
added some additional comments

quinton-hoole commented:
@AmitRoushan @VenuReddy2103 What is the status with this PR please?

quinton-hoole commented:
@AmitRoushan @VenuReddy2103 What is the status with this PR please?
